src/qibolab/result.py

```diff
@@ -1,100 +1,91 @@
-from dataclasses import dataclass
 from functools import cached_property
 from typing import Optional
 
 import numpy as np
 import numpy.typing as npt
 
-ExecRes = np.dtype([("i", np.float64), ("q", np.float64)])
+iq_record_dtype = np.dtype([("i", np.float64), ("q", np.float64)])
 
 
-@dataclass
 class ExecutionResults:
     """Data structure to deal with the output of :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`"""
 
-    array: npt.NDArray[ExecRes]
-    shots: Optional[npt.NDArray[np.uint32]] = None
+    def __init__(self, i: np.ndarray | list, q: np.ndarray | list, state: np.ndarray | list = None):
```

Comment:
Accepting more generic inputs is usually a good idea for public interfaces, because it makes you more flexible.

However, proliferating the input types also leads to redundant code.

Here there are type checks and array conversion, but the same ones might be in some other place as well.
So my recommendation would be to either accepts lists or arrays, but only one of the two.

Since in the end you want to work with arrays, by default it is more convenient to directly request them.
But if the typical use case is to handle lists, and then you have to convert them in many places, then stick to lists input, and do here the conversion.

Reply:
Also consider that arrays in general are more flexible, since they might carry more information:
- list -> array is usually a lossless conversion
- array -> list might be lossy

Reply:
Thanks Alessandro,
Arrays are also a lot faster; I support your proposal to accept only np arrays